Sources/FoundationEssentials/String/StringProtocol+Essentials.swift

```diff
@@ -53,57 +54,101 @@ extension String {
                 }
                 return allASCII ? data : nil
             }
-        default:
-#if FOUNDATION_FRAMEWORK
-            // TODO: Implement data(using:allowLossyConversion:) in Swift
-            return _ns.data(
-                using: encoding.rawValue,
-                allowLossyConversion: allowLossyConversion)
+        case .utf16BigEndian, .utf16LittleEndian, .utf16:
+            let bom: UInt16?
+            let swap: Bool
+            
+            if encoding == .utf16 {
+                swap = false
+                bom = 0xFEFF
+            } else if encoding == .utf16BigEndian {
+#if _endian(little)
+                swap = true
 #else
-            switch encoding {
-            case .utf16BigEndian, .utf16LittleEndian:
-                // This creates a contiguous storage for Data to simply memcpy, the most efficient way to give it bytes.
-                return withUnsafeTemporaryAllocation(of: UInt8.self, capacity: self.utf16.count * 2) { utf16Buffer in
-                    _ = utf16Buffer.initialize(from: UTF16ToDataAdaptor(self.utf16, endianness: Endianness(encoding)!))
-                    defer { utf16Buffer.deinitialize() }
-                    return Data(utf16Buffer)
-                }
-            case .utf16:
+                swap = false
+#endif
+                bom = nil
+            } else if encoding == .utf16LittleEndian {
 #if _endian(little)
-                let data = Data([0xFF, 0xFE])
-                let hostEncoding : String.Encoding = .utf16LittleEndian
+                swap = false
 #else
-                let data = Data([0xFE, 0xFF])
-                let hostEncoding : String.Encoding  = .utf16BigEndian
+                swap = true
 #endif
-                guard let swapped = self.data(using: hostEncoding, allowLossyConversion: allowLossyConversion) else {
-                    return nil
+                bom = nil
+            } else {
+                fatalError("Unreachable")
+            }
+            
+            // Grab this value once, as it requires doing a calculation over String's UTF8 storage
+            let inputCount = self.utf16.count
+            
+            // The output may have 1 additional UTF16 character, if it has a BOM
+            let outputCount = bom == nil ? inputCount : inputCount + 1
+            
+            // Allocate enough memory to hold the UTF16 bytes after conversion. We will pass this off to Data.
+            let utf16Pointer = calloc(outputCount, MemoryLayout<UInt16>.size)!.assumingMemoryBound(to: UInt16.self)
```

Comment:
For what it's worth, this `calloc` call prevents me from building the project without `import Darwin`. Tangentially, the memory is unbound, so you ought to use `bindMemory(...)` or `withMemoryRebound(...)` depending on what binding state `Data` wants.

Reply:
Sorry about that build failure you're seeing! I've just merged https://github.com/apple/swift-foundation/pull/680 to address that issue